hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java

```diff
@@ -21,15 +21,193 @@
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj) {
+        return true;
+      } else if (obj == null || getClass() != obj.getClass()) {
+        return false;
+      }
+      final ByteArray that = (ByteArray) obj;
+      return Arrays.equals(this.bytes, that.bytes);
+    }
+
+    @Override
+    public int hashCode() {
+      return Arrays.hashCode(bytes);
+    }
+  }
+
+  class PutOpCache {
+    class FamilyCache {
+      private final ColumnFamily family;
+      /** A (key -> value) map. */
+      private final Map<ByteArray, byte[]> putOps = new HashMap<>();
+
+      private int batchSize;
+      private int discardedSize;
+      private int discardedCount;
+      private int putCount;
+      private int delCount;
+
+      FamilyCache(ColumnFamily family) {
+        this.family = family;
+      }
+
+      /** Batch put the entire family cache. */
+      void batchPut() throws IOException {
+        for (Map.Entry<ByteArray, byte[]> op : putOps.entrySet()) {
+          family.batchPut(writeBatch, op.getKey().bytes, op.getValue());
+        }
+        putOps.clear();
+      }
+
+      void put(byte[] key, byte[] value) {
+        batchSize += key.length + value.length;
+        putCount++;
+        final byte[] previousValue = putOps.put(new ByteArray(key), value);
+
+        if (previousValue != null) {
+          discardedSize += key.length + previousValue.length;
+          discardedCount++;
+          debug(() -> String.format("Overwriting a previous put(value: %s)",
+              byteSize2String(previousValue.length)));
+        }
+
+        debug(() -> String.format(
+            "%s: %s put(key: %s, value: %s), newPut=%s, %s; key=%s",
+            name, family.getName(), byteSize2String(key.length),
+            byteSize2String(value.length),
+            putCount, getBatchSizeDiscardedString(),
+            StringUtils.bytes2HexString(key).toUpperCase()));
+      }
+
+      void remove(byte[] key) {
+        batchSize += key.length;
+        delCount++;
+        final byte[] removed = putOps.remove(new ByteArray(key));
+        if (removed != null) {
+          discardedSize += key.length + removed.length;
+          discardedCount++;
+          debug(() -> String.format("Removed a previous put(value: %s)",
+              byteSize2String(removed.length)));
+        }
+
+        debug(() -> String.format(
+            "%s: %s delete(key: %s), newDel=%s, %s; key=%s",
+            name, family.getName(), byteSize2String(key.length),
+            delCount, getBatchSizeDiscardedString(),
+            StringUtils.bytes2HexString(key).toUpperCase()));
+      }
+
+      String getBatchSizeDiscardedString() {
+        return String.format("batchSize=%s, discarded: %s",
+            byteSize2String(batchSize),
+            countSize2String(discardedCount, discardedSize));
+      }
+    }
+
+    /** A (family name -> {@link FamilyCache}) map. */
+    private final Map<String, FamilyCache> map = new HashMap<>();
```

Comment:
I suggest renaming the variable to nameToCacheMap for easier understanding.

Reply:
Sure.